File: apps/web/src/features/routing/stopNames.test.ts
import { describe, it, expect, vi } from 'vitest';
import type { LatLng } from '../../geo';
import { createStopNames } from './stopNames';

const at = (lat: number, lng: number): LatLng => ({ lat, lng });

/** Lazy, cached reverse-geocoding of route stops: a name is fetched at most once
 *  per ~11 m grid cell, survives a reorder (it's keyed on the coordinate), and a
 *  failed/empty lookup never throws — the caller falls back to trimmed coords. */
describe('stop name resolution', () => {
  it('resolves a stop to its geocoded name', async () => {
    const names = createStopNames(async () => 'Storgata 1');
    expect(await names.resolve(at(69.96, 23.27))).toBe('Storgata 1');
  });

  it('geocodes a coordinate only once, then serves the cache on a re-render', async () => {
    const geocode = vi.fn(async () => 'Storgata 1');
    const names = createStopNames(geocode);
    const p = at(69.96, 23.27);
    await names.resolve(p);
    await names.resolve(p);
    expect(geocode).toHaveBeenCalledTimes(1);
    expect(names.cached(p)).toBe('Storgata 1');
  });

  it('keeps the name for a stop after a reorder moves it (same coordinate, later query)', async () => {
    const geocode = vi.fn(async () => 'Fjelltoppen');
    const names = createStopNames(geocode);
    const p = at(69.9607, 23.2715);
    await names.resolve(p); // resolved while it was, say, the 2nd stop
    // after a reorder the row for the same place asks again — served from cache,
    // even for a coordinate a few metres off inside the same ~11 m cell
    expect(names.cached(at(69.96072, 23.27152))).toBe('Fjelltoppen');
    expect(geocode).toHaveBeenCalledTimes(1);
  });

  it('yields undefined and never throws when the lookup fails (offline)', async () => {
    const names = createStopNames(async () => {
      throw new Error('offline');
    });
    await expect(names.resolve(at(69.96, 23.27))).resolves.toBeUndefined();
  });

  it('does not cache an empty result, so a later online attempt can still succeed', async () => {
    let online = false;
    const names = createStopNames(async () => (online ? 'Storgata 1' : ''));
    const p = at(69.96, 23.27);
    expect(await names.resolve(p)).toBeUndefined();
    online = true;
    expect(await names.resolve(p)).toBe('Storgata 1');
  });
});

File: apps/android/feature/map/src/test/kotlin/com/sigmundgranaas/turbo/expressive/feature/map/StopNamesTest.kt
package com.sigmundgranaas.turbo.expressive.feature.map

import com.sigmundgranaas.turbo.expressive.core.common.Outcome
import com.sigmundgranaas.turbo.expressive.core.data.ReverseGeocodeRepository
import com.sigmundgranaas.turbo.expressive.domain.LatLng
import com.sigmundgranaas.turbo.expressive.domain.LocationDescription
import com.sigmundgranaas.turbo.expressive.domain.PlaceQualifier
import com.sigmundgranaas.turbo.expressive.feature.map.route.StopLabels
import com.sigmundgranaas.turbo.expressive.feature.map.route.StopNames
import com.sigmundgranaas.turbo.expressive.feature.map.route.StopPalette
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertNull
import org.junit.Test

/** A scriptable reverse-geocoder: returns [next] as the stop's title (or fails), counting calls. */
private class FakeGeocoder(var next: LocationDescription?) : ReverseGeocodeRepository {
    var calls = 0
    override suspend fun describe(point: LatLng): Outcome<LocationDescription> {
        calls++
        return next?.let { Outcome.Success(it) } ?: Outcome.Failure(IllegalStateException("offline"))
    }
}

private fun desc(title: String) = LocationDescription(title, PlaceQualifier.On)

class StopNamesTest {

    private val p = LatLng(69.9607, 23.2715)

    @Test
    fun `trimmed coords are plain comma-separated decimals`() {
        assertEquals("69.9607, 23.2715", StopLabels.trimmedCoords(p))
    }

    @Test
    fun `label shows the cached name when present, else the trimmed coords`() {
        assertEquals("Besseggen", StopLabels.label("Besseggen", p))
        assertEquals(StopLabels.trimmedCoords(p), StopLabels.label(null, p))
        assertEquals(StopLabels.trimmedCoords(p), StopLabels.label("  ", p)) // blank name ignored
    }

    @Test
    fun `resolve returns the geocoded name when online`() = runTest {
        val names = StopNames(FakeGeocoder(desc("Galdhøpiggen")))
        assertEquals("Galdhøpiggen", names.resolve(p))
    }

    @Test
    fun `offline resolve yields null and never throws (row falls back to coords)`() = runTest {
        val names = StopNames(FakeGeocoder(next = null))
        assertNull(names.resolve(p))
        // The display rule then shows coordinates — no exception, no blank row.
        assertEquals(StopLabels.trimmedCoords(p), StopLabels.label(names.cached(p), p))
    }

    @Test
    fun `a resolved name is cached — a re-render does not geocode again`() = runTest {
        val geo = FakeGeocoder(desc("Besseggen"))
        val names = StopNames(geo)
        assertEquals("Besseggen", names.resolve(p))
        assertEquals(1, geo.calls)
        // The place "changes name" upstream, but the cached value stands (no second fetch).
        geo.next = desc("Somewhere else")
        assertEquals("Besseggen", names.resolve(p))
        assertEquals("no second geocode for the same cell", 1, geo.calls)
        assertEquals("Besseggen", names.cached(p))
    }

    @Test
    fun `a via keeps its palette colour regardless of list position`() {
        // Colour is keyed on the coordinate's grid cell, so reordering the stops around it
        // never changes the colour the user sees for that stop.
        val before = StopPalette.colorOf(p)
        val after = StopPalette.colorOf(LatLng(p.lat, p.lng)) // same place, different object
        assertEquals(before, after)
    }

    @Test
    fun `distinct places generally get distinct palette colours`() {
        val c1 = StopPalette.colorOf(LatLng(69.0, 18.0))
        val c2 = StopPalette.colorOf(LatLng(60.0, 10.0))
        assertNotEquals(c1, c2)
    }
}
